File: frontend/src/app/services/topicPresentation.ts
import presentation from '../config/topicPresentation.json';

type TopicPresentation = {
  groups: {
    order: string[];
    ru: Record<string, string>;
  };
  categoryToGroup: Record<string, string>;
  categoryRu: Record<string, string>;
  topicRu: Record<string, string>;
};

const topicPresentation = presentation as TopicPresentation;

export const TOPICS_PAGE_GROUPS_EN = topicPresentation.groups.order;
export const TOPICS_PAGE_GROUPS_RU = TOPICS_PAGE_GROUPS_EN.map(
  (group) => topicPresentation.groups.ru[group] || group,
);
export const CATEGORY_RU = topicPresentation.categoryRu;
export const TOPIC_RU = topicPresentation.topicRu;

function normalizeLabel(value: any): string {
  if (typeof value === 'string') return value.replace(/\s+/g, ' ').trim();
  if (value === null || value === undefined) return '';
  return String(value).replace(/\s+/g, ' ').trim();
}

export function groupTopicCategoryForTopicsPage(categoryRaw: any): string {
  const category = normalizeLabel(categoryRaw);
  if (!category) return 'Admin';
  return topicPresentation.categoryToGroup[category] || 'Admin';
}

export function translateTopicRu(topicRaw: any): string {
  const topic = normalizeLabel(topicRaw);
  if (!topic) return topic;
  return TOPIC_RU[topic] || topic;
}

export function translateCategory(categoryRaw: any, ru = false): string {
  const category = normalizeLabel(categoryRaw);
  if (!ru) return category || 'General';
  return CATEGORY_RU[category] || category || 'Общий';
}

export function translateTopicsPageGroup(groupRaw: any, ru = false): string {
  const group = normalizeLabel(groupRaw);
  if (!ru) return group;
  return topicPresentation.groups.ru[group] || group;
}

File: utils/topic_presentation.py
from __future__ import annotations

"""Shared presentation metadata for Topics-page grouping and Russian labels."""

from functools import lru_cache
import json
from pathlib import Path
from typing import Any


_TOPIC_PRESENTATION_PATH = (
    Path(__file__).resolve().parents[1]
    / "frontend"
    / "src"
    / "app"
    / "config"
    / "topicPresentation.json"
)


@lru_cache(maxsize=1)
def _load_topic_presentation() -> dict[str, Any]:
    with _TOPIC_PRESENTATION_PATH.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def topic_group_for_category(category: str) -> str:
    name = str(category or "").strip()
    if not name:
        return "Admin"
    payload = _load_topic_presentation()
    return str(payload.get("categoryToGroup", {}).get(name, "Admin"))


def category_ru(category: str) -> str:
    name = str(category or "").strip()
    if not name:
        return name
    payload = _load_topic_presentation()
    return str(payload.get("categoryRu", {}).get(name, name))


def topic_ru(topic: str) -> str:
    name = str(topic or "").strip()
    if not name:
        return name
    payload = _load_topic_presentation()
    return str(payload.get("topicRu", {}).get(name, name))


def topics_page_groups_en() -> list[str]:
    payload = _load_topic_presentation()
    groups = payload.get("groups", {})
    return [str(item) for item in groups.get("order", [])]

